gems/pending/disk/modules/miq_disk_cache.rb

```diff
@@ -0,0 +1,109 @@
+    @dInfo.down_stream      = downstream
+
+    MiqDisk.new(self, @dInfo, 0)
+  end
+
+  def d_init
+    @block_cache      = LruHash.new(@dInfo.cache_size)
+    @cache_hits       = Hash.new(0)
+    @cache_misses     = Hash.new(0)
+    @blockSize        = @dInfo.block_size
+    @downstream       = @dInfo.down_stream
+    @sectors_to_cache = @dInfo.sectors_to_cache
+  end
+
+  def d_size
+    @d_size ||= @downstream.d_size
+  end
+
+  def d_read(pos, len)
+    $log.debug "MiqDiskCache.d_read(#{pos}, #{len})"
+    return nil if pos >= d_size
+    len = d_size - pos if (pos + len) > d_size
+
+    start_sector, start_offset = pos.divmod(@blockSize)
+    end_sector                 = (pos + len - 1) / @blockSize
+    number_sectors             = end_sector - start_sector + 1
+ 
+    d_read_cached(start_sector, number_sectors)[start_offset, len]
+  end
+
+  def d_read_cached(start_sector, number_sectors)
+    $log.debug "MiqDiskCache.d_read_cached(#{start_sector}, #{number_sectors})"
+    @block_cache.keys.each do |block_range|
+      sector_offset = start_sector - block_range.first
+      buffer_offset = sector_offset * @blockSize
+      if block_range.include?(start_sector) && block_range.include?(start_sector + number_sectors - 1)
+        length = number_sectors * @blockSize
+        @cache_hits[start_sector] += 1
+        return @block_cache[block_range][buffer_offset, length]
+      elsif block_range.include?(start_sector)
+        sectors_in_range = block_range.last - start_sector
+        length           = sectors_in_range * @blockSize
+        remaining_blocks = number_sectors - sectors_in_range
+        @cache_hits[start_sector] += 1
+        return @block_cache[block_range][buffer_offset, length] + d_read_cached(block_range.last + 1, remaining_blocks)
+      elsif block_range.include?(start_sector + number_sectors - 1)
+        sectors_in_range = (start_sector + number_sectors) - block_range.first
+        length           = sectors_in_range * @blockSize
+        remaining_blocks = number_sectors - sectors_in_range
+        @cache_hits[start_sector] += 1
+        return d_read_cached(start_sector, remaining_blocks) + @block_cache[block_range][block_range.first, length]
+      end
+    end
+    block_range               = entry_range(start_sector, number_sectors)
+    range_length              = (block_range.last - block_range.first + 1) * @blockSize
+    @block_cache[block_range] = @downstream.d_read(block_range.first * @blockSize, range_length)
+    @cache_misses[start_sector] += 1
+
+    sector_offset             = start_sector - block_range.first
+    buffer_offset             = sector_offset * @blockSize
+    length                    = number_sectors * @blockSize
+
+    @block_cache[block_range][buffer_offset, length]
+  end
+
+  def d_close
+    hit_or_miss if DEBUG_CACHE_STATS
+    @downstream.d_close
+  end
+
+  def method_missing(m, *args)
+    @downstream.send(m, *args)
+  end
+
+  private
+
+  def hit_or_miss
```

Comment:
ok, if it's for consistency, then leave for now, and we can change them all at once.